### PR TITLE
add CMFGEN photoionization cross-sections

### PR DESCRIPTION
**Pull request by epassaro**
**Description**

This long and ugly PR add photoionization cross sections from the CMFGEN database to Carsus.

**Motivation and context**

**How has this been tested?**
- [x] Testing pipeline.
- [ ] Other. 

**Examples**

- [Cross sections](https://gist.github.com/epassaro/2302761ffa98ef804a8494344849f10a)
- [Create atom data](https://gist.github.com/epassaro/2676ab19fe0ae0f58a2f3895a6d82db8)

**Type of change**

- [ ] Bug fix. 
- [x] New feature. 
- [x] Breaking change. 
- [ ] None of the above. 

**Checklist**

- [x] My change requires a change to the documentation.
    - [x] I have updated the documentation accordingly.
    - [x] (optional) I have built the documentation on my fork following [the instructions](https://tardis-sn.github.io/tardis/development/documentation_preview.html).
- [x] I have assigned and requested two reviewers for this pull request.

### CODE REVIEW

File: carsus/io/cmfgen/base.py [hunk 1]
```diff
@@ -657,8 +564,216 @@ def __init__(self, data, priority=10):
...
+            ion_levels['label'] = ion_levels['label'].str.split('[', expand=True)
+
+            try:
+                match = ion_levels.set_index('label').loc[[lower_level_label]]
+
+            except KeyError:
+                logger.warning(f'Level not found: \'{lower_level_label}\'.')
+                continue
+
+            lambda_angstrom = match['Lam(A)'].tolist()
```
Comment (afloers) [inline carsus/io/cmfgen/base.py:665]: we should never use `Lam(A)` - this column has the smallest number of digits. Instead use `E(cm^-1)`. 

Also, the energy required to ionise from any level is given by ionisation energy (`!Ionization energy` in the header of the file) minus the level energy (from column `E(cm^-1)`.

File: carsus/io/cmfgen/base.py [hunk 2]
```diff
@@ -657,8 +564,216 @@ def __init__(self, data, priority=10):
...
+            level_number = (match['ID'] -1).tolist()
+
+            # Get statistical weights for J-splitted levels
+            match['w'] = match['g']/match.sum()['g']
+            w = match['w'].tolist()
+
+            # match is > 1 just for J-splitted levels
+            for j in range(len(match)):
+                threshold_energy_ryd = HC_IN_EV_ANGSTROM / lambda_angstrom[j] / RYD_TO_EV
```
Comment (afloers) [inline carsus/io/cmfgen/base.py:674]: see comment above - this is not the threshold energy. Instead do `ionisation_energy - level_energy`.

Comment (afloers) [inline carsus/io/cmfgen/base.py:674]: Why is this in angstrom? I find it quite difficult to follow these unit conversions.

File: carsus/io/cmfgen/base.py [hunk 3]
```diff
@@ -698,10 +819,40 @@ def _get_levels_lines(self, data):
...
+                    hyd = reader['hyd'].apply(lambda x: 10 ** (8 + x))
+                    gbf = reader['gbf']
+
+                    hyd_phixs, hyd_phixs_energy_grid_ryd = {}, {}
+                    hyd_gaunt_factor, hyd_gaunt_energy_grid_ryd = {}, {}
+
+                    for n in range(1, n_levels+1):
+
+                        lambda_angstrom = lvl.loc[n-1, 'Lam(A)']
+                        e_threshold_ev = HC_IN_EV_ANGSTROM / lambda_angstrom
```
Comment (afloers) [inline carsus/io/cmfgen/base.py:838]: see comment above about threshold energy

File: carsus/io/cmfgen/base.py (hunk 2 above)
Comment (epassaro) [reply carsus/io/cmfgen/base.py:674]: This was extracted from Luke's code: https://github.com/artis-mcrt/artisatomic/blob/4959ba96c71a6a81d395a9eada259ae99bb1ab4a/artisatomic/readhillierdata.py#L625

File: carsus/io/cmfgen/base.py [hunk 4]
```diff
@@ -698,10 +819,40 @@ def _get_levels_lines(self, data):
...
+
+                        hyd_gaunt_energy_grid_ryd[n] = [e_threshold_ev / RYD_TO_EV * 10 ** (n_start_u + n_del_u * i) for i in range(n_points)]
```
Comment (afloers) [inline carsus/io/cmfgen/base.py:840]: with line 838: HC_IN_EV_ANGSTROM / RYD_TO_EV
we should remove unnecessary conversions

File: carsus/io/cmfgen/util.py [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+
+    # Only applies to `He`. The threshold cross sections seems ok, but 
+    # energy dependence could be slightly wrong. What is the `h` parameter
+    # that is not used?.
+    
+    energy_grid = np.linspace(0.0, 1.0, n_points, endpoint=False)
+    phixs_table = np.empty((len(energy_grid), 2))
+
+    for i, c in enumerate(energy_grid):
+        energy_div_threshold = 1 + 20 * (c ** 2)
```
Comment (chvogl) [inline carsus/io/cmfgen/util.py:266]:
```suggestion
    for i, c1 in enumerate(energy_grid):
        energy_div_threshold = 1 + 20 * (c1 ** 2)
```

Comment (chvogl) [reply carsus/io/cmfgen/util.py:266]: Otherwise this overwrites the fit coefficient c yielding incorrect results.

File: carsus/io/cmfgen/util.py [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+
+        Journal of Physics B: Atomic, Molecular, and Optical Physics, 
+        Volume 21, Issue 22, pp. 3669-3683 (1988).
+    """
+    energy_grid = np.linspace(0.0, 1.0, n_points, endpoint=False)
+    phixs_table = np.empty((len(energy_grid), 2))
+
+    for i, c in enumerate(energy_grid):
+
+        energy_div_threshold = 1 + 20 * (c ** 2)
```
Comment (chvogl) [inline carsus/io/cmfgen/util.py:233]:
```suggestion
    for i, c1 in enumerate(energy_grid):

        energy_div_threshold = 1 + 20 * (c1 ** 2)
```

File: carsus/io/cmfgen/util.py [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+        J. Quant. Spectrosc. Radiat. Transfer. Vol 12, pp. 299-306.
+    """
+    
+    raise NotImplementedError
+
+
+def get_null_phixs_table(n_points=1000):
+    """
+    """
+    phixs_table = np.column_stack((np.linspace(0, 1.0, 1000, endpoint=False),
```
Comment (chvogl) [inline carsus/io/cmfgen/util.py:329]: This is not particularly crucial but it would be nice if the energies start at the threshold energy (as for the other cross-section types) and not at 0.

File: carsus/io/cmfgen/util.py (hunk 5 above)
Comment (epassaro) [reply carsus/io/cmfgen/util.py]: Done. I used the same name convention from `artisatomic`.

File: carsus/data/config.yml [hunk 8]
```diff
@@ -0,0 +1,130 @@
+atom:
+  H: 
+    name: HYD
```
Comment (andrewfullard) [inline carsus/data/config.yml:3]: Comment on why HYD

File: carsus/data/config.yml [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    ion_charge:
+      0:
+        date: 5dec96
+        osc: hi_osc.dat
```
Comment (andrewfullard) [inline carsus/data/config.yml:7]: Add comment # oscillator strength

File: carsus/data/config.yml [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+        hyd: hyd_l_data.dat  # hydrogen log cross-sections
+        gbf: gbf_n_data.dat  # bound-free gaunt factors
+        pho:
```
Comment (andrewfullard) [inline carsus/data/config.yml:10]: Add comment # photoionization cross-sections

File: carsus/data/config.yml (hunk 8 above)
Comment (andrewfullard) [reply carsus/data/config.yml:3]: Description should be in docs

File: carsus/io/cmfgen/util.py [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+import astropy.constants as const
+import astropy.units as u
+import numpy as np
+
+RYD_TO_EV = u.rydberg.to('eV')
+H_IN_EV_SECONDS = const.h.to('eV s').value
+HC_IN_EV_ANGSTROM = (const.h * const.c).to('eV angstrom').value
+
+
+def open_cmfgen_file(fname, encoding='ISO-8859-1'):
```
Comment (andrewfullard) [inline carsus/io/cmfgen/util.py:13]: Change to ASCII? Check why ISO

File: carsus/io/cmfgen/util.py [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+    -------
+    float
+    """
+    try:
+        value = float(string.replace('D', 'E'))
+
+    except ValueError:
+
+        # Typo in `MG/VIII/23oct02/phot_sm_3000`, line 23340
+        if string == '1-.00':
```
Comment (andrewfullard) [inline carsus/io/cmfgen/util.py:35]: Place above try/except (both ifs)

File: carsus/io/cmfgen/util.py [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+            value = 10.00
+
+        # Typo in `SUL/V/08jul99/phot_op.big`, lines 9255-9257
+        if string == '*********':
```
Comment (andrewfullard) [inline carsus/io/cmfgen/util.py:39]: Ask Hillier about the real number

File: carsus/io/cmfgen/util.py [hunk 14]
```diff
@@ -0,0 +1,332 @@
...
+                if string1 in line and string2 not in line:
+                    break
+
+        else:
+            n, line = None, None
+
+    if row_number is True:
+        return n
+
+    return line
```
Comment (andrewfullard) [inline carsus/io/cmfgen/util.py:96]:
```suggestion
    return line
```
returning strings or ints?

Comment (andrewfullard) [inline carsus/io/cmfgen/util.py:96]:
```suggestion
    return n, line
```
Return both to avoid confusion on if function is returning int or string

Comment (andrewfullard) [reply carsus/io/cmfgen/util.py:96]: Can be unclear if function is returning different types depending on input

File: carsus/io/cmfgen/util.py [hunk 15]
```diff
@@ -0,0 +1,332 @@
...
+        for line in itertools.islice(f, start, stop):
+            for k in keys:
+                if k.lower() in line.lower():
+                    meta[k.strip('!')] = line.split()[0]
+
+    return meta
+
+
+def get_seaton_phixs_table(threshold_energy_ryd, sigma_t, beta, s, nu_0=None, n_points=1000):
+    """
```
Comment (andrewfullard) [inline carsus/io/cmfgen/util.py:131]: Missing

File: carsus/io/cmfgen/util.py [hunk 16]
```diff
@@ -0,0 +1,332 @@
...
+    """
+    energy_grid = np.linspace(0.0, 1.0, n_points, endpoint=False)
+    phixs_table = np.empty((len(energy_grid), 2))
+
+    for i, c in enumerate(energy_grid):
+        energy_div_threshold = 1 + 20 * (c ** 2)
```
Comment (andrewfullard) [inline carsus/io/cmfgen/util.py:137]: Why these numbers?

File: carsus/io/cmfgen/util.py (hunk 15 above)
Comment (andrewfullard) [reply carsus/io/cmfgen/util.py:131]: Cite source paper

File: carsus/io/cmfgen/util.py [hunk 17]
```diff
@@ -0,0 +1,332 @@
...
+            else:
+                cross_section = 0.0
+
+        phixs_table[i] = energy_div_threshold * threshold_energy_ryd, cross_section
+
+    return phixs_table
+
+
+def get_hydrogenic_n_phixs_table(hyd_gaunt_energy_grid_ryd, hyd_gaunt_factor, threshold_energy_ryd, n):
+    """
```
Comment (andrewfullard) [inline carsus/io/cmfgen/util.py:161]: Missing docstring, cite

File: carsus/io/cmfgen/util.py [hunk 18]
```diff
@@ -0,0 +1,332 @@
...
+    """
+    energy_grid = hyd_gaunt_energy_grid_ryd[n]
+    phixs_table = np.empty((len(energy_grid), 2))
+    scale_factor = 7.91 / threshold_energy_ryd / n
```
Comment (andrewfullard) [inline carsus/io/cmfgen/util.py:165]: Why?

File: carsus/io/cmfgen/util.py [hunk 19]
```diff
@@ -0,0 +1,332 @@
...
+        phixs_table[i][1] = cross_section
+
+    return phixs_table
+
+
+def get_hydrogenic_nl_phixs_table(hyd_phixs_energy_grid_ryd, hyd_phixs, threshold_energy_ryd, n, l_start, l_end, nu_0=None):
+    """
+    """
+
+    assert l_start >= 0
```
Comment (andrewfullard) [inline carsus/io/cmfgen/util.py:185]: Raise exception?

Comment (andrewfullard) [reply carsus/io/cmfgen/util.py:185]: Add descriptive output string

File: carsus/io/cmfgen/util.py [hunk 20]
```diff
@@ -0,0 +1,332 @@
...
+    assert l_end <= n - 1
+
+    energy_grid = hyd_phixs_energy_grid_ryd[(n, l_start)]
+    phixs_table = np.empty((len(energy_grid), 2))
+
+    threshold_energy_ev = threshold_energy_ryd * RYD_TO_EV
+    scale_factor = 1 / threshold_energy_ryd / (n ** 2) / ((l_end - l_start + 1) * (l_end + l_start + 1))
```
Comment (andrewfullard) [inline carsus/io/cmfgen/util.py:192]: Simplify

File: carsus/io/cmfgen/util.py [hunk 21]
```diff
@@ -0,0 +1,332 @@
...
+
+    for i, energy_ryd in enumerate(energy_grid):
+        energy_div_threshold = energy_ryd / energy_grid[0]
+        if nu_0 is None:
+            U = energy_div_threshold
```
Comment (andrewfullard) [inline carsus/io/cmfgen/util.py:197]: Is this really a constant if it changes depending on if?

File: carsus/io/cmfgen/util.py [hunk 22]
```diff
@@ -0,0 +1,332 @@
...
+        Leibowitz, E. M.
+
+        J. Quant. Spectrosc. Radiat. Transfer. Vol 12, pp. 299-306.
+    """
+    
+    raise NotImplementedError
+
+
+def get_null_phixs_table(n_points=1000):
+    """
```
Comment (andrewfullard) [inline carsus/io/cmfgen/util.py:327]: Missing docstring

File: carsus/io/output/base.py [hunk 23]
```diff
@@ -71,9 +71,12 @@ def __init__(self,
         self.create_macro_atom()
         self.create_macro_atom_references()
 
-        if (chianti_reader is not None) and (not chianti_reader.collisions.empty):
+        if (chianti_reader is not None) and (chianti_reader.collisions.empty):
```
Comment (andrewfullard) [inline carsus/io/output/base.py:74]: Why is this changed?

File: carsus/io/output/base.py [hunk 24]
```diff
@@ -629,6 +632,19 @@ def create_collisions(self, temperatures=np.arange(2000, 50000, 2000)):
 
         return collisions
 
+    def create_cross_sections(self):
+        """
```
Comment (andrewfullard) [inline carsus/io/output/base.py:636]: Missing docstring

File: carsus/io/output/base.py [hunk 25]
```diff
@@ -629,6 +632,19 @@ def create_collisions(self, temperatures=np.arange(2000, 50000, 2000)):
...
+        """
+        logger.info('Ingesting photoionization cross-sections.')
+
+        cross_sections = self.cmfgen_reader.cross_sections
+        cross_sections['energy'] = u.Quantity(cross_sections['energy'], u.Unit('Ry')).to('eV').to(u.Hz, equivalencies=u.spectral()).value
+        cross_sections['sigma'] = cross_sections['sigma']*1E-18  # Megabarns to cm^2
```
Comment (andrewfullard) [inline carsus/io/output/base.py:642]: units

File: carsus/io/output/base.py [hunk 26]
```diff
@@ -629,6 +632,19 @@ def create_collisions(self, temperatures=np.arange(2000, 50000, 2000)):
 
         return collisions
 
+    def create_cross_sections(self):
+        """
+        """
+        logger.info('Ingesting photoionization cross-sections.')
+
+        cross_sections = self.cmfgen_reader.cross_sections
+        cross_sections['energy'] = u.Quantity(cross_sections['energy'], u.Unit('Ry')).to('eV').to(u.Hz, equivalencies=u.spectral()).value
```
Comment (andrewfullard) [inline carsus/io/output/base.py:641]: Why extra steps between Ry and Hz?
```suggestion
        cross_sections['energy'] = u.Quantity(cross_sections['energy'], u.Unit('Ry')).to(u.Hz, equivalencies=u.spectral()).value
```

File: carsus/io/cmfgen/base.py [hunk 27]
```diff
@@ -146,49 +41,45 @@ class CMFGENEnergyLevelsParser(BaseParser):
...
-            df = pd.read_csv(fname, **kwargs, engine='python')
+            df = pd.read_csv(fname, **config)
 
         except pd.errors.EmptyDataError:
             df = pd.DataFrame(columns=columns)
-            logger.warn(f'Table is empty: `{fname}`.')
+            logger.warning(f'Empty table: `{fname}`.')
 
-        # Assign column names by file content
         if df.shape[1] == 10:
```
Comment (andrewfullard) [inline carsus/io/cmfgen/base.py:61]: Why 10?

File: carsus/io/cmfgen/base.py [hunk 28]
```diff
@@ -146,49 +41,45 @@ class CMFGENEnergyLevelsParser(BaseParser):
...
-            # Read column names and split them keeping one space (e.g. '10^15 Hz')
+            # Read column names and split them keeping just one space (e.g. '10^15 Hz')
```
Comment (andrewfullard) [inline carsus/io/cmfgen/base.py:62]: REVIEW STOPPED HERE CONTINUE FROM HERE

File: carsus/io/cmfgen/util.py [hunk 29]
```diff
@@ -0,0 +1,339 @@
...
+        energy_div_threshold = 1 + 20 * (c ** 2)
+
+        if nu_0 is None:
+            threshold_div_energy = energy_div_threshold ** -1
+            cross_section = sigma_t * (beta + (1 - beta) * threshold_div_energy) * (threshold_div_energy ** s)
+
+        else:
+            threshold_energy_ev = threshold_energy_ryd * RYD_TO_EV
+            energy_offset_div_threshold = energy_div_threshold + (nu_0 * 1e15 * H_IN_EV_SECONDS) / threshold_energy_ev
+            threshold_div_energy_offset = energy_offset_div_threshold ** -1
```
Comment (chvogl) [inline carsus/io/cmfgen/util.py:152]:
```suggestion
            threshold_div_energy_offset = energy_div_threshold**-1 * (1 + (nu_0 * 1e15 * H_IN_EV_SECONDS) / threshold_energy_ev)
```
I think the offset should be added to the ionization edge instead of the current energy. See the [CMFGEN source code](https://github.com/juliotux/cmfgen/blob/6cae33c174a561fe26ec9fb13efd55345b5fe4e1/newsubs/sub_phot_gen.f#L482).

Comment (lukeshingles) [reply carsus/io/cmfgen/util.py:152]: Thanks for finding this. I suggest renaming threshold_div_energy_offset to offset_threshold_div_energy just to be slightly more clear.

Comment (chvogl) [reply carsus/io/cmfgen/util.py:152]: That's a good point. I have changed it.

File: carsus/io/cmfgen/util.py [hunk 30]
```diff
@@ -0,0 +1,339 @@
...
+
+            if threshold_div_energy_offset < 1.0:
+                cross_section = sigma_t * (beta + (1 - beta) * (threshold_div_energy_offset)) * \
+                    (threshold_div_energy_offset ** s)
```
Comment (chvogl) [inline carsus/io/cmfgen/util.py:156]:
```suggestion
            if offset_threshold_div_energy < 1.0:
                cross_section = sigma_t * (beta + (1 - beta) * (offset_threshold_div_energy)) * \
                    (offset_threshold_div_energy ** s)
```
See comment above by @lukeshingles.

File: carsus/data/config.yml [hunk 31]
```diff
@@ -0,0 +1,130 @@
+atom:
+  H:
+    name: HYD                # atom name in CMFGEN database
```
[the remark this answers is not in the document]
Comment (epassaro) [reply carsus/data/config.yml]: I realized that's exactly what we were doing and the key `name` is never used. I removed it.

File: carsus/io/cmfgen/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (epassaro) [reply carsus/io/cmfgen/base.py:858]: `HC_IN_EV_ANGSTROM` is already expressed with AstroPy units!.

File: carsus/io/output/base.py (hunk 25 above)
Comment (epassaro) [reply carsus/io/output/base.py]: Done.

File: carsus/io/cmfgen/util.py (hunk 15 above)
Comment (epassaro) [reply carsus/io/cmfgen/util.py:129]: Done.

File: carsus/io/cmfgen/util.py (hunk 16 above)
Comment (epassaro) [reply carsus/io/cmfgen/util.py]: Done.

File: carsus/io/cmfgen/util.py [hunk 32]
```diff
@@ -0,0 +1,332 @@
...
+            cross_section = 0.0
+
+        phixs_table[i][0] = energy_div_threshold * threshold_energy_ryd
+        phixs_table[i][1] = cross_section
+
+    return phixs_table
+
+
+def get_hydrogenic_nl_phixs_table(hyd_phixs_energy_grid_ryd, hyd_phixs, threshold_energy_ryd, n, l_start, l_end, nu_0=None):
+    """
```
[the remark this answers is not in the document]
Comment (epassaro) [reply carsus/io/cmfgen/util.py:188]: Can't find

File: carsus/io/cmfgen/util.py (hunk 19 above)
Comment (epassaro) [reply carsus/io/cmfgen/util.py:210]: I don't know if these assertions are useful, probably not.

File: carsus/io/cmfgen/base.py [hunk 33]
```diff
@@ -657,8 +564,216 @@ def __init__(self, data, priority=10):
...
+            w = match['w'].tolist()
+
+            # match is > 1 just for J-splitted levels
+            for j in range(len(match)):
+                threshold_energy_ryd = HC_IN_EV_ANGSTROM / lambda_angstrom[j] / RYD_TO_EV
+
+                if cross_section_type == 0:
+                    phixs_table = get_null_phixs_table()
+
+                elif cross_section_type in [20, 21, 22]:
```
[the remark this answers is not in the document]
Comment (epassaro) [reply carsus/io/cmfgen/base.py]: Done

File: carsus/io/cmfgen/base.py [hunk 34]
```diff
@@ -698,10 +735,48 @@ def _get_levels_lines(self, data):
...
+                        hyd_gaunt_energy_grid_ryd[n] = [e_threshold_ev / RYD_TO_EV * 10 ** (n_start_u + n_del_u * i) for i in range(n_points)]
+                        hyd_gaunt_factor[n] = gbf.loc[n].tolist()
+
+                        for l in range(0, n):
+                            hyd_phixs_energy_grid_ryd[(n, l)] = [e_threshold_ev / RYD_TO_EV * 10 ** (l_start_u + l_del_u * i) for i in range(l_points)]
+                            hyd_phixs[(n,l)] = hyd.loc[(n,l)].tolist()
+
+                pxs = self.cross_sections_squeeze(reader['cross_sections'][0], lvl, hyd_phixs_energy_grid_ryd, hyd_phixs, hyd_gaunt_energy_grid_ryd, hyd_gaunt_factor)
+                pxs['atomic_number'] = ion[0]
+                pxs['ion_charge'] = ion[1]
```
Comment (chvogl) [inline carsus/io/cmfgen/base.py:773]:
```suggestion
                pxs['ion_number'] = ion[1]
```
To be compatible with TARDIS.

File: carsus/io/cmfgen/base.py [hunk 35]
```diff
@@ -719,6 +794,17 @@ def _get_levels_lines(self, data):
...
                        'gf', 'j_lower', 'j_upper', 'wavelength']]
 
+        if 'ionization_energy' in reader.keys():
+            ionization_energies = pd.DataFrame.from_records(ioz_list)
+            ionization_energies['ionization_energy'] = (ionization_energies['ionization_energy'].values / u.cm).to('eV', equivalencies=u.spectral()).value
+            ionization_energies = ionization_energies.set_index(['atomic_number', 'ion_charge']).squeeze()
+            self.ionization_energies = ionization_energies
+
+        if 'cross_sections' in reader.keys():
+            cross_sections = pd.concat(pxs_list)
```
Comment (chvogl) [inline carsus/io/cmfgen/base.py:804]:
```suggestion
            cross_sections = pd.concat(pxs_list)
            cross_sections.rename({'ion_charge': 'ion_number', 'label': 'level_number'})
```
To make it compatible with TARDIS.

File: carsus/io/output/base.py [hunk 36]
```diff
@@ -629,6 +656,26 @@ def create_collisions(self, temperatures=np.arange(2000, 50000, 2000)):
...
+        pandas.DataFrame
+
+        """
+
+        logger.info('Ingesting photoionization cross-sections.')
+
+        cross_sections = self.cmfgen_reader.cross_sections
+        cross_sections['energy'] = u.Quantity(cross_sections['energy'], 'Ry').to('Hz', equivalencies=u.spectral())
+        cross_sections['sigma'] = u.Quantity(cross_sections['sigma'], 'Mbarn').to('cm2')
+        cross_sections.columns = ['nu', 'x_sect']
```
Comment (Rodot-) [inline carsus/io/output/base.py:674]:
```suggestion
        cross_sections = cross_sections.rename({'energy':'nu', 'sigma':'x_sect'}, axis=1)
```

File: carsus/io/cmfgen/base.py [hunk 37]
```diff
@@ -719,6 +794,21 @@ def _get_levels_lines(self, data):
...
+            cross_sections = cross_sections.rename(
+                {'ion_charge': 'ion_number', 'level_index': 'level_number'},
+                axis=1
+            )
+            cross_sections = cross_sections.set_index(['atomic_number', 'ion_charge', 'level_index'])
```
Comment (chvogl) [inline carsus/io/cmfgen/base.py:809]:
```suggestion
            cross_sections = cross_sections.set_index(['atomic_number', 'ion_number', 'level_number'])
```

File: carsus/io/cmfgen/base.py (hunk 35 above)
Comment (epassaro) [reply carsus/io/cmfgen/base.py:804]: Dropped the commit, this should be done in the output module (see https://github.com/tardis-sn/carsus/issues/214)

File: carsus/io/output/base.py (hunk 36 above)
Comment (epassaro) [reply carsus/io/output/base.py]: Done.

Comment (epassaro) [reply carsus/io/output/base.py]: Same as previous comment.

File: carsus/io/cmfgen/base.py (hunk 37 above)
Comment (epassaro) [reply carsus/io/cmfgen/base.py:805]: Same as https://github.com/tardis-sn/carsus/pull/251#discussion_r747621067

File: carsus/io/output/base.py [hunk 38]
```diff
@@ -816,6 +863,22 @@ def collisions_prepared(self):
...
+        Prepare the DataFrame with photoionization cross-sections for TARDIS.
+
+        Returns
+        -------
+        pandas.DataFrame
+
+        """
+        cross_sections_prepared = self.cross_sections.reset_index()
+        cross_sections_prepared = cross_sections_prepared.rename(columns=({'ion_charge': 'ion_number'}))
+        cross_sections_prepared = cross_sections_prepared.set_index(['atomic_number', 'ion_number', 'level_index'])
```
Comment (chvogl) [inline carsus/io/output/base.py:878]:
```suggestion
        cross_sections_prepared = cross_sections_prepared.set_index(['atomic_number', 'ion_number', 'level_number'])
```

File: carsus/io/output/base.py [hunk 39]
```diff
@@ -678,8 +667,17 @@ def create_cross_sections(self):
...
 
         logger.info('Ingesting photoionization cross-sections.')
+        cross_sections_reader = self.cmfgen_reader.cross_sections
+
+        # remove cross-sections above the ionization threshold
+        phixs_clean = []
+        for ion in self.cmfgen_ions:
+            ion_phixs = cross_sections_reader.loc[ion]
+            threshold = self.ionization_energies.base[ion]
+            ion_phixs = ion_phixs[ ion_phixs['energy'] < threshold ]
```
Comment (chvogl) [inline carsus/io/output/base.py:677]: Does this not also remove the high energy part of the cross section table for levels below the ionization threshold?
Say we have an ionization threshold of 10 eV and a level with an excitation energy of 1 eV; In this case the table would start at 9 eV and then be incorrectly truncated at 10 eV?

Comment (chvogl) [reply carsus/io/output/base.py:677]: On second thought, I don't think this removes anything that we want to be remove.

Comment (epassaro) [reply carsus/io/output/base.py:677]: Yes, it's clear for me now
